plugins/formal-verification/src/org/jetbrains/kotlin/formver/embeddings/ExpEmbedding.kt

```diff
@@ -0,0 +1,187 @@
+    override val type: TypeEmbedding = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.Mod =
+        Exp.Mod(left.withType(type).viperExp, right.withType(type).viperExp)
+}
+
+data class LtCmp(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    val cmpArgType = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.LtCmp =
+        Exp.LtCmp(left.withType(cmpArgType).viperExp, right.withType(cmpArgType).viperExp)
+}
+
+data class LeCmp(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    val cmpArgType = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.LeCmp =
+        Exp.LeCmp(left.withType(cmpArgType).viperExp, right.withType(cmpArgType).viperExp)
+}
+
+data class GtCmp(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    val cmpArgType = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.GtCmp =
+        Exp.GtCmp(left.withType(cmpArgType).viperExp, right.withType(cmpArgType).viperExp)
+}
+
+data class GeCmp(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    val cmpArgType = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.GeCmp =
+        Exp.GeCmp(left.withType(cmpArgType).viperExp, right.withType(cmpArgType).viperExp)
+}
+
+data class EqCmp(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    val cmpArgType = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.EqCmp =
+        Exp.EqCmp(left.withType(cmpArgType).viperExp, right.withType(cmpArgType).viperExp)
+}
+
+data class NeCmp(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    val cmpArgType = commonSuperType(left.type, right.type)
+    override val viperExp: Exp.NeCmp =
+        Exp.NeCmp(left.withType(cmpArgType).viperExp, right.withType(cmpArgType).viperExp)
+}
+
+data class And(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    override val viperExp: Exp.And =
+        Exp.And(left.viperExp, right.viperExp)
+}
+
+data class Implies(val left: ExpEmbedding, val right: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    override val viperExp: Exp.Implies =
+        Exp.Implies(left.viperExp, right.viperExp)
+}
+
+data class Not(val arg: ExpEmbedding) : ExpEmbedding {
+    override val type: TypeEmbedding = BooleanTypeEmbedding
+    override val viperExp: Exp.Not =
+        Exp.Not(arg.viperExp)
+}
+
+data class Trigger(val exps: List<ExpEmbedding>) {
```

Comment:
Does it make sense to have `Trigger`, `Exists`, and `Forall` at the `ExpEmbedding` level?  There is no Kotlin construct that they embed...